bugsnag-android-core/src/main/java/com/bugsnag/android/ReportDeliveryDelegate.java

```diff
@@ -0,0 +1,109 @@
+package com.bugsnag.android;
+
+import androidx.annotation.NonNull;
+
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.RejectedExecutionException;
+
+class ReportDeliveryDelegate extends BaseObservable {
+
+    final Logger logger;
+    final EventStore eventStore;
+    final ImmutableConfig immutableConfig;
+    final BreadcrumbState breadcrumbState;
+
+    ReportDeliveryDelegate(Logger logger, EventStore eventStore,
+                           ImmutableConfig immutableConfig, BreadcrumbState breadcrumbState) {
+        this.logger = logger;
+        this.eventStore = eventStore;
+        this.immutableConfig = immutableConfig;
+        this.breadcrumbState = breadcrumbState;
+    }
+
+    void deliverEvent(@NonNull Event event) {
+        // Build the report
+        Report report = new Report(immutableConfig.getApiKey(), event);
+        Session session = event.getSession();
+
+        if (session != null) {
+            setChanged();
+
+            if (event.isUnhandled()) {
+                event.setSession(session.incrementUnhandledAndCopy());
+                notifyObservers(StateEvent.NotifyUnhandled.INSTANCE);
+            } else {
+                event.setSession(session.incrementHandledAndCopy());
+                notifyObservers(StateEvent.NotifyHandled.INSTANCE);
+            }
+        }
+
+        if (event.isUnhandled()) {
+            eventStore.write(event);
+            eventStore.flushAsync();
```

Comment:
This feels like the wrong level of abstraction and gives this method too many things to know.

For example, consider extracting to a `cacheEvent` function which takes a boolean for whether to flush straightaway. This can be called by all the methods that use `eventStore` in the class.

Reply:
Extracted to a separate function.